File: Networking/Objects/NetworkVariableSyncMessage.cs
using Networking.Data;

namespace Networking.Objects
{
    public struct NetworkVariableSyncMessage : IMessage
    {
        public ushort typeHash;
        public ushort hash;

        public IMessage msg;

        public NetworkVariableSyncMessage(ushort typeHash, ushort hash, IMessage msg)
        {
            this.typeHash = typeHash;
            this.hash = hash;
            this.msg = msg;
        }

        public void Deserialize(Reader reader)
        {
            typeHash = reader.ReadUShort();
            hash = reader.ReadUShort();
            msg = reader.ReadAnonymous() as IMessage;
        }

        public void Serialize(Writer writer)
        {
            writer.WriteUShort(typeHash);
            writer.WriteUShort(hash);
            writer.WriteAnonymous(msg);
        }
    }
}

File: Networking/Objects/NetworkVariableSyncMessage.vb
Imports Common.Extensions

Imports Networking.Data

Namespace Networking.Objects
    Public Structure NetworkVariableSyncMessage
        Implements IMessage
        Public objectId As Integer

        Public hash As UShort

        Public msg As IMessage

        Public Sub New(objectId As Integer, hash As UShort, msg As IMessage)
            Me.objectId = objectId
            Me.hash = hash
            Me.msg = msg
        End Sub

        Public Sub Deserialize(reader As Reader) Implements IDeserialize.Deserialize
            objectId = reader.ReadInt()
            hash = reader.ReadUShort()

            msg = TryCast(reader.ReadType().Construct(), IMessage)
            msg.Deserialize(reader)
        End Sub

        Public Sub Serialize(writer As Writer) Implements ISerialize.Serialize
            writer.WriteInt(objectId)
            writer.WriteUShort(hash)

            msg.Serialize(writer)
        End Sub
    End Structure
End Namespace
